### ISSUE TITLE
Error with AssertionConsumerServiceURL using OAuth0

### ISSUE THREAD

**Issue by kewinbrand** [nextcloud/user_saml#343]
Hi, I'm trying to setup nextcloud using auth0 as Identity Provider.

The plugin correctly redirects me to the external login page but then an generic error from oauth0 is displayed. In the oauth0 logs it says:

"error": {
      "message": "The SAML Request AssertionConsumerServiceURL is invalid: 'http://my-nextcloud-domain/apps/user_saml/saml/acs'",
      "oauthError": "invalid_request",
      "type": "request-error"
    }

I inspected the redirects that the browser does when loggin in and indeed the SAMLRequest that nextcloud passes in the assertion service url w/o https. Whole SAMLRequest:

<samlp:AuthnRequest
    xmlns:samlp="urn:oasis:names:tc:SAML:2.0:protocol"
    xmlns:saml="urn:oasis:names:tc:SAML:2.0:assertion"
    ID="ONELOGIN_edc7929f6bf546831d8b2bce06b64bc9a3ccec0d"
    Version="2.0"

    IssueInstant="2019-07-11T15:23:32Z"
    Destination="-omitted-"
    ProtocolBinding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST"
    AssertionConsumerServiceURL="http://my-nextcloud-domain/apps/user_saml/saml/acs">
    <saml:Issuer>http://my-nextcloud-domain/apps/user_saml/saml/metadata</saml:Issuer>
    <samlp:NameIDPolicy
        Format="urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified"
        AllowCreate="true" />

</samlp:AuthnRequest>

My suspicion is that as the nextcloud is behind an reverse proxy (nginx) it does not use https altough it is.

Can you shed some light?
Btw I followed this guide: https://medium.com/@mathiasconradt/nextcloud-single-sign-on-with-auth0-a546cdf1fccf

**Comment by xtrasimplicity** [comment]
I'm having the same issue, and it appears that we're not alone.
https://github.com/nextcloud/user_saml/issues/227

This was working for me with NC 14, but our upgrade to 16 has broken it.

**Comment by xtrasimplicity** [comment]
After a little digging, I can confirm that the URLs are built using OC's `URLGenerator`, which respects the `HTTP_X_FORWARDED_PROTO` header and `overwriteprotocol` configuration option, as per: https://github.com/nextcloud/server/blob/23245904d386144090b313568b09385a88291e73/lib/private/AppFramework/Http/Request.php#L688

TL;DR: You can resolve this by making sure that the proxy server sets the `HTTP_X_FORWARDED_PROTO` header to `https`, or by setting `overwriteprotocol` to `https` in `/var/www/html/config/config.php`, like so:
```php
<?php
$CONFIG = array (
 ....
  "overwriteprotocol" => "https",
);
```


Source: https://docs.nextcloud.com/server/10/admin_manual/configuration_server/reverse_proxy_configuration.html#multiple-domains-reverse-ssl-proxy

**Comment by danysvirsky** [comment]
In my case it was the incorrect "Allowed Callback URLs", I used the guide and it says to set
"https://<NEXTCLOUD_HOSTNAME>/index.php/apps/user_saml/saml/acs"
Try setting it to
"https://<NEXTCLOUD_HOSTNAME>/apps/user_saml/saml/acs"

**Comment by hexxone** [comment]
> Had the same issue. In my case the problem was incorrect "Allowed Callback URLs", I used the same guide and it says to set "https://<NEXTCLOUD_HOSTNAME>/index.php/apps/user_saml/saml/acs" Try setting it to "https://<NEXTCLOUD_HOSTNAME>/apps/user_saml/saml/acs"

I dont understand this problem.
I also keep having the same one still in 2025, 6 years later.

It seems like its pure RANDOM which of the urls nextcloud expects?
Because it seems to me like it changes every 2 weeks.. exactly then I had to change it from `/index.php/apps/` to just `/apps/` and now the exact OPPOSITE is the case and it wants the index.php back?

What is going on here?

```
Bad Request
ACS URL of 
https://cloud.test.com/index.php/apps/user_saml/saml/acs 
doesn't match Provider ACS URL of 
https://cloud.test.com/apps/user_saml/saml/acs.
```

**Comment by hexxone** [comment]
> forget it! my mistake!
> the URL was changed by the pretty-URL configuration in Nextcloud!
> https://docs.nextcloud.com/server/latest/admin_manual/installation/source_installation.html#pretty-urls

Yes, I also have that setting on enabled.
However it still doesn't explain and make sense why that changes on every update?

Or did I miss something?

**Comment by joern15** [comment]
> Yes, I also have that setting on enabled.
> However it still doesn't explain and make sense why that changes on every update?
> 
> Or did I miss something?
> 

in my case, for some reason the .htaccess file was missing after an update.
 After recreating it with the command
`occ maintenance:update:htaccess`
the URLs were pretty again and also the AssertionConsumerServiceURL was set without /index.php.

**Comment by blizzz** [comment]
pretty/non-pretty url config is not in the scope of the SAML backend. I close the issue hence as configuration topic. If you have any follow ups about the config, please check the issues on the server repository instead.

**Comment by susnux** [comment]
@blizzz but apps need to always work with `index.php` URLs. Meaning even when pretty URLs are enabled the ´index.php/` URL variant needs to work as well as there are cases where this is expected even from server

**Comment by blizzz** [comment]
> [@blizzz](https://github.com/blizzz) but apps need to always work with `index.php` URLs. Meaning even when pretty URLs are enabled the ´index.php/` URL variant needs to work as well as there are cases where this is expected even from server

The SAML protocol is strict in that regard and you can have only one URL to your service provider (i.e. Nextcloud here). Otherwise the validation fails.
